mdata/store_cassandra.go

```diff
@@ -73,13 +76,61 @@ It's safe for concurrent use by multiple goroutines and a typical usage scenario
 object to interact with the whole Cassandra cluster.
 */
 
+type ttlTables map[uint32]struct {
+	table  string
+	window uint32
+}
+
 type cassandraStore struct {
 	session     *gocql.Session
 	writeQueues []chan *ChunkWriteRequest
 	readQueue   chan *ChunkReadRequest
+	ttlTables   ttlTables
+}
+
+func ttlUnits(ttl uint32) uint32 {
+	// convert ttl to hours
+	return ttl / (60 * 60)
+}
+
+func getTtlTables(ttls []uint32, windowFactor int, nameFormat string) ttlTables {
+	tables := make(ttlTables)
+	for _, ttl := range ttls {
+		/*
+		 * the purpose of this loop is to bucket metrics of similar TTLs.
+		 * we first calculate the largest power of 2 that's smaller than the TTL and then divide the result by
+		 * the window factor. for example with a window factor of 20 we want to group the metrics like this:
+		 *
+		 * |--------------------|--------------------|-------------------|
+		 * |          TTL range |         Table Name |       window size |
+		 * |--------------------|--------------------|-------------------|
+		 * |                  1 |          metrics_1 |                 1 |
+		 * |              2 - 3 |          metrics_2 |                 1 |
+		 * |              4 - 7 |          metrics_4 |                 1 |
+		 * |             8 - 15 |          metrics_8 |                 1 |
+		 * |            16 - 31 |         metrics_16 |                 1 |
+		 * |            32 - 63 |         metrics_32 |                 2 |
+		 * |           64 - 127 |         metrics_64 |                 4 |
+		 * |          128 - 255 |        metrics_128 |                 7 |
+		 * |          256 - 511 |        metrics_256 |                13 |
+		 * |         512 - 1023 |        metrics_512 |                26 |
+		 * |        1024 - 2047 |       metrics_1024 |                52 |
+		 * |--------------------|--------------------|-------------------|
+		 *
+		 */
+
+		// calculate the window size by finding the largest power of 2 that's smaller than ttl
+		window := uint32(math.Exp2(math.Floor(math.Log2(float64(ttlUnits(ttl))))))
+		tableName := fmt.Sprintf(nameFormat, window)
+		result := tables[ttl]
```

Comment:
Yeah I just did that because the alternative would be pretty clunky, because that's an anonymous struct:
```
                tables[ttl] = struct {
                        table  string
                        window uint32
                }{
                        table:  tableName,
                        window: window/uint32(windowFactor) + 1,
                }
```